### PR TITLE
smoothing

### PR DESCRIPTION
**Pull request by cxhernandez**
ref https://github.com/mdtraj/mdtraj/issues/957

### CODE REVIEW

File: mdtraj/core/trajectory.py [hunk 1]
```diff
@@ -1669,6 +1669,74 @@ def remove_solvent(self, exclude=None, inplace=False):
...
+        inplace : bool, default=False
+            The return value is either ``self``, or the new trajectory,
+            depending on the value of ``inplace``.
+        Returns
+        -------
+        traj : md.Trajectory
+            The return value is either ``self``, or the new smoothed trajectory,
+            depending on the value of ``inplace``.
+        """
+        from scipy.signal import lfilter, lfilter_zi, filtfilt, butter
```
Comment (cxhernandez) [inline mdtraj/core/trajectory.py]: Wasn't sure if I should import these at the top...

**Comment by mpharrigan** [comment]
smooth like butter...worth

File: mdtraj/core/trajectory.py [hunk 2]
```diff
@@ -1669,6 +1669,74 @@ def remove_solvent(self, exclude=None, inplace=False):
 
         return self.atom_slice(atom_indices, inplace = inplace)
 
+    def smooth(self, width=11, order=3, atom_indices=None, inplace=False):
+        """
+        Smoothen a trajectory using a zero-delay Buttersworth filter.
+        This code is copied from the scipy cookbook, with sytlistic improvements.
+        http://www.scipy.org/Cookbook/FiltFilt
+        Parameters
```
Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: linebreak before parameters

File: mdtraj/core/trajectory.py [hunk 3]
```diff
@@ -1669,6 +1669,74 @@ def remove_solvent(self, exclude=None, inplace=False):
...
+        order : int, optional
+            The order of the filter. A small odd number is recommended. Higher
+            order filters cutoff more quickly, but have worse numerical
+            properties.
+        atom_indices : array-like, dtype=int, shape=(n_atoms)
+            List of indices of atoms to retain in the new trajectory.
+        inplace : bool, default=False
+            The return value is either ``self``, or the new trajectory,
+            depending on the value of ``inplace``.
+        Returns
```
Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: likewise

**Comment by cxhernandez** [comment]
@mpharrigan you should give this a try on some of your trajectories. I've only tried it on WW, but the results looked pretty good.

**Comment by mpharrigan** [comment]
Should we include this in v1.5 cc #959 ?

**Comment by cxhernandez** [comment]
I think it'd be nice to have in v1.5

**Comment by cxhernandez** [comment]
Any idea what the appveyor errors are for py3?

**Comment by mpharrigan** [comment]
well the build matrices are different, which leads me to believe that the "push" ci uses older code from the rest of the codebase whereas the "pr" ci does a merge with HEAD first

**Comment by mpharrigan** [comment]
You should probably add a test for this function

File: mdtraj/core/trajectory.py [hunk 4]
```diff
@@ -1669,6 +1669,76 @@ def remove_solvent(self, exclude=None, inplace=False):
...
 
+    def smooth(self, width=11, order=3, atom_indices=None, inplace=False):
+        """
+        Smoothen a trajectory using a zero-delay Buttersworth filter.
+        This code is copied from the scipy cookbook, with sytlistic improvements.
+        http://www.scipy.org/Cookbook/FiltFilt
+        
+        Parameters
+        ----------
+        width : float
```
Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: you have a check below that makes sure this is int

Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: I don't think `width` should have a default

File: mdtraj/core/trajectory.py [hunk 5]
```diff
@@ -1669,6 +1669,76 @@ def remove_solvent(self, exclude=None, inplace=False):
...
+            This acts very similar to the window_len in the window smoother. In
+            the implementation, the frequency of the low-pass filter is taken to
+            be two over this width, so it's like "half the period" of the sinusiod
+            where the filter starts to kick in.
+        order : int, optional
+            The order of the filter. A small odd number is recommended. Higher
+            order filters cutoff more quickly, but have worse numerical
+            properties.
+        atom_indices : array-like, dtype=int, shape=(n_atoms)
+            List of indices of atoms to retain in the new trajectory.
```
Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: Add something about the default is None and that keeps all the atoms

File: mdtraj/core/trajectory.py [hunk 6]
```diff
@@ -1669,6 +1669,76 @@ def remove_solvent(self, exclude=None, inplace=False):
...
+
+        xyz = self.xyz.copy()
+
+        for i in atom_indices:
+            for j in range(3):
+
+                signal = xyz[:, i, j]
+
+                # find nearest odd integer
+                pad = int(np.ceil((width + 1)/2)*2 - 1)
```
Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: Move outside the loops?

Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: Actually almost all of this should be moved outside the loop, no?

Comment (cxhernandez) [inline mdtraj/core/trajectory.py]: good catch!

Comment (rmcgibbo) [inline mdtraj/core/trajectory.py]: do they scipy filter functions allow broadcasting, or is it really necessary to loop over the atom indices and dimensions.

File: mdtraj/core/trajectory.py (hunk 4 above)
Comment (cxhernandez) [inline mdtraj/core/trajectory.py]: Yeah that was left over from @rmcgibbo's code. He has this line that converts it into an odd integer:

`pad = int(np.ceil((width + 1)/2)*2 - 1)`

Philosophically, I think the user should know that `width` has to be an odd whole number so they have at least some idea of what's going on under the hood, but maybe that's asking for too much?

Comment (mpharrigan) [inline mdtraj/core/trajectory.py]: Whatever you think is best re: int or float, just make sure the docstring is correct

I don't think it should have a default value though

**Comment by cxhernandez** [comment]
What should the test look like? Making sure it spits out a trajectory of the same length?

**Comment by hainm** [comment]
> What should the test look like? Making sure it spits out a trajectory of the same length?

What are your expected traj? assert to your expected one. You can visualize it, if you think the smoothing seems ok, I think you can assert to that (or to output from other program).

**Comment by mpharrigan** [comment]
This needs a good alignment within the trajectory right?

File: mdtraj/core/trajectory.py [hunk 7]
```diff
@@ -1669,6 +1669,80 @@ def remove_solvent(self, exclude=None, inplace=False):
 
         return self.atom_slice(atom_indices, inplace = inplace)
 
+    def smooth(self, width, order=3, atom_indices=None, inplace=False):
+        """
```
Comment (rmcgibbo) [inline mdtraj/core/trajectory.py]: No need for the extra linebreak here, and put the reference to the scipy cookbook as a citation instead (e.g. https://github.com/mdtraj/mdtraj/blob/master/mdtraj/geometry/dssp.py#L75-L78).

**Comment by rmcgibbo** [comment]
Yeah, I think it would might make sense to either do a progressive alignment of the frames before smoothing, or at least to mention something about alignment in the docs.

**Comment by cxhernandez** [comment]
Yep, alignment is necessary.

**Comment by cxhernandez** [comment]
Alright. I added added frame0-smooth.dcd which is used as a reference for the `test_smooth`.

**Comment by cxhernandez** [comment]
One of the appveyor tests failed on the trajectory comparison :confused:

**Comment by cxhernandez** [comment]
Any ideas why the trajectories are different for Win/Lin?

**Comment by cxhernandez** [comment]
Any ideas why the trajectories are different for Win/Lin?

**Comment by rmcgibbo** [comment]
Strange.

**Comment by rmcgibbo** [comment]
Can you try adding a simpler test case that applies the same filter function to a known 1D signal, like `np.sin(np.arange(100))` or something?

**Comment by rmcgibbo** [comment]
e.g.

```
pad = 5
signal = np.sin(np.arange(100))
padded = np.r_[signal[pad - 1: 0: -1], signal, signal[-1: -pad: -1]]
z, _ = lfilter(b, a, padded, zi=zi*padded[0])
[... etc]

assert result == [hardcoded result that you compute on your computer and just copy-paste in]
```
And then we can see if it matches in windows.

**Comment by rmcgibbo** [comment]
I can also hop on a Windows VM, or boot into Windows in lab.

**Comment by cxhernandez** [comment]
I just tried it on a VM, and the results are not the same :disappointed:

**Comment by rmcgibbo** [comment]
What exactly isn't consistent? Can you get it down to a single method?

**Comment by cxhernandez** [comment]
Actually, even generating the signal using `np.sin(np.arange(100))` results in small differences.

**Comment by rmcgibbo** [comment]
Are you testing for actual equality? https://possiblywrong.wordpress.com/2013/11/15/floating-point-equality-its-worse-than-you-think/

**Comment by cxhernandez** [comment]
Okay– after using `mdtraj.testing.eq`, everything seems to check out.

**Comment by rmcgibbo** [comment]
Sweet!

**Comment by rmcgibbo** [comment]
FWIW, I would recommend reading http://www.johndcook.com/blog/2009/04/06/numbers-are-a-leaky-abstraction/

**Comment by rmcgibbo** [comment]
Is this still a work in progress? It looks ready to merge to me.

**Comment by rmcgibbo** [comment]
p.s. Thanks for implementing this!

**Comment by cxhernandez** [comment]
np!

**Pull request merged**